File: source/c/alloc_defer_free.c
// gcc -o alloc_defer_free alloc_defer_free.c
#include <stdio.h>
#include <stdlib.h>

void demo(void) {
    int *a = malloc(8);
    int *b = malloc(8);
    int *c = malloc(8);
    // defer free(c); defer free(b); defer free(a); → free at end in reverse order

    *a = 10; *b = 20; *c = 30;

    int va = *a, vb = *b, vc = *c;
    printf("%d %d %d\n", va, vb, vc);
    printf("all memory freed\n");

    free(c); free(b); free(a);
}

int main(void) {
    demo();
    return 0;
}

File: source/rust/alloc_defer_free.rs
use std::alloc::{alloc, dealloc, Layout};

fn main() {
    let layout = Layout::from_size_align(8, 4).unwrap();
    unsafe {
        let a = alloc(layout);
        let b = alloc(layout);
        let c = alloc(layout);

        *(a as *mut i32) = 10;
        *(b as *mut i32) = 20;
        *(c as *mut i32) = 30;

        let va = *(a as *const i32);
        let vb = *(b as *const i32);
        let vc = *(c as *const i32);

        println!("{} {} {}", va, vb, vc);
        println!("all memory freed");

        dealloc(a, layout);
        dealloc(b, layout);
        dealloc(c, layout);
    }
}
